### ISSUE TITLE
Define `DataPlane` status conditions specification

### ISSUE THREAD

**Issue by pmalek** [Kong/kong-operator#143]
### Problem statement

Currently, `DataPlane` receives updates  from the operator setting only 1 status condition: `Ready`.

Its semantics are not obvious to users and have to be defined.

### Proposal

- Answer the following questions (this list might not be exhaustive at this point and might be extended in the future):
  
  - When the underlying `Deployment` is configured via e.g. `spec.deployment.replicas` to have N replicas, when should the `DataPlane` have its `status.conditions` updated with `Ready` condition, with `status` set to `True`?
    - When N replicas are ready?
    - When N replicas are available?
    - When `Deployment`'s `status.observedGeneration == metadata.generation`?
    - When `Deployment`'s status conditions contain `Progressing` condition with status set to `True` and reason set to `NewReplicaSetAvailable`. NOTE: `Deployment` status conditions do not have `ObservedGeneration` field.
    - When all or some of the above circumstances are met?
  - During a rollout (by default using a `RollingUpdate` strategy), when the `Deployment` is updated old, ready Pods can still serve traffic, should the `DataPlane` be marked as `Ready` with `status=True` or `status=False`?
  - Users would like to be able to determine if the new spec has been correctly propagated and all the `DataPlane` Pods are running that up to date spec. `Ready` can do one of 2 things:
    - Report that the update has concluded and every managed `DataPlane` has been updated. Before that happens, `Ready` status condition has `status=False`. This will report `DataPlane` as not ready during an update (current status in 1.1)
    - Report that the `DataPlane` is reachable (taking into account still yet to be updated Pods) and every managed `DataPlane` has been updated. This will make `Ready` status condition to have `status=True` during an update but will preclude the user from being able to know when the `DataPlane` has reached its eventually consistent state.

- Update operator code if the answers are different than what's currently implemented
- Write answers to those questions in a doc, preferably in the one created as an artifact of https://github.com/Kong/gateway-operator/issues/1315

### Acceptance criteria

- [ ] As a user I know what `Ready` status condition mean `DataPlane` 
- [ ] As a user I am able to deduce from `DataPlane` state when it has concluded an update after applying an update to its spec.

**Comment by mheap** [comment]
I think we need to expand the statuses on `DataPlane` here. Here's my gut feel for what I'd expect as a user:

* `Ready`: The DataPlane is ready to handle traffic (either on the old pods, or on new ones).
* `RolloutComplete`: `False` for initial deployment
    * When using `RollingUpdate`: `False` until all pods are using the new spec
    * When using `BlueGreen`: `False` until `promote-when-ready` is applied + the new replicas take over
    * When using `Canary` (future): `False` until all pods are using the new spec

If people want to know if they can send traffic, use `Ready`. If they want to know if the latest spec is being used, check `RolloutComplete`.

I've opted to keep `Ready` as the "reachable" state so that people aren't surprised if they run `k get pods xyx` and see `Ready: 0/4` after changing a deployment spec.

**Comment by pmalek** [comment]
That makes sense I believe. Just to make it a bit more precise:

> * `Ready`: The DataPlane is ready to handle traffic (either on the old pods, or on new ones).

What does it mean "to be able to handle traffic"? If I want N (where N>1) replicas, should Ready be set with 1 or precisely with N ready replicas? The latter could potentially cause the status to flap when 1 replica would get killed or unresponsive for whatever reason.

**Comment by mheap** [comment]
Good question. I think where N > 1 replicas (unless we want to add a minimum threshold such as 50% of replicas are ready)

**Comment by shaneutt** [comment]
Consider a theoretical scenario: I'm doing some multi-cluster or even geo-loadbalancing with Kong. Let's say I'm bringing up, or maybe rolling out updates to one of my cluster's `DataPlanes`, which I've enabled autoscaling for with a minimum of 8 replicas, because I know as soon as I add this cluster to the LB it's going to receive an amount of traffic that requires at least 8 replicas to comfortably handle. If `Ready` means N > 1 and I use this as the trigger to enable balancing to this cluster, perhaps I start sending all traffic to only a couple of replicas, and they fall over. The problem is similar for 50% replicas.

The above kind of scale isn't necessarily the most common deployment, but I don't think we want to be in a state where such a user effectively has to traverse owner relations in order to get better data for actuating workflows. "ready" has a bit of an unfortunate history in these kinds of systems (in fact, we struggled with it in Gateway API a bit too, right up to the v1 release), and I think it's important to remind us that `Deployment` provides `readyReplicas` and `availableReplicas` so that users of `Deployment` can "decide for themselves".

I'm open to being persuaded... but after thinking this over my inclination is that we should be exposing `readyReplicas` and `availableReplicas` so that `DataPlane` API users can "decide for themselves" if they like to, but then `ready` should just be a shorthand of these meaning "all replicas are _ready_ (specifically)". This seems to cover all needs, and the default position is "everything you specified you wanted is complete, ready, and has been ready for X period" as opposed to ready meaning "some of what you wanted is complete". Let me know your thoughts :thinking:

**Comment by mheap** [comment]
The "it falls over if I send all my traffic" use case was on my mind too.

I'm ok with `Ready` meaning that all replicas are ready

**Comment by sentinelleader** [comment]
agree to the above points, now being using kgo on a live environment in cgw, we look for signals 1) did KGO applied the spec fully 2) Are all pods are ready with the new spec that was applied. If one of them is failing we consider ourselves as degraded because the desired state requested by the customer is not yet achieved. I think that is the guarantee that ive seen with manby of the operators like helm controller for example

**Comment by pmalek** [comment]
@shaneutt I agree but I'd like to be precise on some of the aspects of what you brought up:

- replicas being available: I fear that this might be misleading when users would be looking at a `Ready` condition and get the "available" replicas count and not the "ready" ones. That can obviously be explained in the docs that are yet to be created as part of https://github.com/Kong/gateway-operator/issues/1315 but I believe it's worth pointing out.

- [we already expose](https://github.com/Kong/gateway-operator/blob/25184da5dd2d77511d10d043c7292dde36770c4a/apis/v1beta1/dataplane_types.go#L157-L165) `ReadyReplicas` and `Replicas` in `DataPlane` status so extending this with `AvailableReplicas` will be rather trivial

- > but then ready should just be a shorthand of these meaning "all replicas are available (specifically)".
  > "everything you specified you wanted is complete, ready, and has been ready for X period (available)"
  
  When user updates e.g. an image, in the middle of an update, the above would imply that  `Ready` condition should flip to `False`, right?

  This way a single condition should be enough.


---

Just to note that we already have the `status.rollout.conditions` that extend the status API for the BG rollout (which is orthogonal to the `Deployment` `RollingUpgrade` strategy that we use for our managed `Deployment`)

**Comment by shaneutt** [comment]
> replicas being available: I fear that this might be misleading

No you're right, I misspoke: I meant _ready_ not _available_, as in the probes are successful. I fixed that in the above comment to avoid confusion.

> extending this with AvailableReplicas will be rather trivial

:+1:

> When user updates e.g. an image, in the middle of an update, the above would imply that Ready condition should flip to False, right?

This is a great question, because it digs deeper down into "well how far do we actually go" and also points out the general problem with composite/summary statuses like "ready" because there's too much to summarize.

Flipping to false when a workflow is underway does definitely have some downsides: it could end up being odd in some cases: ~if a load-balancer were configured to use the `ready` state to determine health, that might be pretty crap and we would have to instruct them to take a different approach~ if someone was integrating with the ready status it might not be intuitive if it was ready=false even though the replicas were all active and traffic was otherwise serve-able.

I wrote down some potential options while I was thinking about this:

1. remove `ready` and leave it up to the user to compile the other status elements together to suit their own needs
2. document the current `ready` field to make it clear that it will go false during workflows (e.g. upgrades), even if the minReplicas will be satisfied during the duration of those workflows
3. replace `ready` with something else, possibly multiple "summary" status fields to indicate different things (e.g. maybe one for "minimum replicas are ready and available to serve traffic (even if the image is being upgraded, transition should be graceful)", and maybe one for "workflows are underway").
4. define (and document) ready as meaning ONLY `readyReplicas >= minReplicas`

And then I had some coffee and thought them over a little bit. They all have downsides: 1 and 3 for example kinda scream "new API version" which is a pretty big "no thanks", and 2 just continues to feel too unintuitive to me the more I think about it. So I think I'm leaning toward option 4, which doesn't feel _perfectly_ intuitive but in the documentation for the field we can try to add strongly worded clarity about what it does NOT cover, and maybe provide the specific example that if the API caller wants to know if workflows are underway, there's no summary field for that for now they just need to go inspect the `status.rollout.conditions` and `status.conditions` as you suggested.

:thinking: let me know your thoughts?

**Comment by sentinelleader** [comment]
> if a load-balancer were configured to use the ready state to determine health, that might be pretty crap and we would have to instruct them to take a different approach

@shaneutt why would a load balancer rely on the kgo crd status, shouldnt it be always relying on service -> endpoints -> selector labels -> only read pods added to endpoints ?

**Comment by shaneutt** [comment]
Fair, too contrived of an example. My point was that if someone was integrating with the ready status it might not be intuitive if it was ready=false even though the replicas were all active and traffic was otherwise serve-able.

**Comment by pmalek** [comment]
@shaneutt 

1. > define (and document) ready as meaning ONLY `readyReplicas` >= `minReplicas`

    `minReplicas` only exists in autoscaling with HPA context. When setting `replicas` on a `DataPlane` we would not have `minReplicas` available or set anywhere, right?

1. `status.rollout.conditions` is only filled for BG rollout enabled `DataPlane`s.

**Comment by shaneutt** [comment]
> minReplicas only exists in autoscaling with HPA context. When setting replicas on a DataPlane we would not have minReplicas available or set anywhere, right?

Right, so `minReplicas` when applicable, or `replicas` when applicable.

> status.rollout.conditions is only filled for BG rollout enabled DataPlanes.

Yes, which is why I also mentioned `status.conditions`. That is to say we should document as part of ready, since it might be the first place someone looks something like "check these other areas to understand if there are changes underway, even if we're otherwise `ready`".

What do you think? Simple enough? Still too confusing?

**Comment by pmalek** [comment]
> Right, so `minReplicas` when applicable, or `replicas` when applicable.

If we go the route: ready means `readyReplicas` >= `replicas` then the status of that condition **will flip** during an update. Just checked with a plain `Deployment` with 5 replicas and these are the statuses that I've observed:

```
        "readyReplicas": 5,
        "replicas": 5,
        "updatedReplicas": 5
...
        "readyReplicas": 5,
        "replicas": 5,
        "unavailableReplicas": 2
...
        "readyReplicas": 4, # Flip to false
        "replicas": 4,
        "unavailableReplicas": 3
...
        "readyReplicas": 4,
        "replicas": 7,
        "unavailableReplicas": 3,
        "updatedReplicas": 3
...
        "readyReplicas": 5, # Flip to true
        "replicas": 7,
        "unavailableReplicas": 2,
        "updatedReplicas": 3
...
        "readyReplicas": 6,
        "replicas": 7,
        "unavailableReplicas": 1,
        "updatedReplicas": 3
```

Which makes sense that Kubernetes takes it down to 4 (from 5) because default strategy is

```
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
```

So we would be more strict and therefore "flip" during update (depending on the `maxUnavailable`).

I'm thinking if using the `Available` status condition from the `Deployment` wouldn't make sense:

```
            {
                "lastTransitionTime": "2024-01-11T20:56:45Z",
                "lastUpdateTime": "2024-01-11T20:56:45Z",
                "message": "Deployment has minimum availability.",
                "reason": "MinimumReplicasAvailable",
                "status": "True",
                "type": "Available"
            },
            {
                "lastTransitionTime": "2024-01-11T19:22:45Z",
                "lastUpdateTime": "2024-01-11T21:07:22Z",
                "message": "ReplicaSet \"kong-gateway-7c65c869b8\" has successfully progressed.",
                "reason": "NewReplicaSetAvailable",
                "status": "True",
                "type": "Progressing"
            }
```

NOTE: We don't expose [deployment strategy](https://pkg.go.dev/k8s.io/api@v0.29.0/apps/v1#DeploymentStrategy) currently in the DP spec so users would have no control over the default that we use.

**Comment by pmalek** [comment]
One more thing I believe it's important to point out: if the `Ready` status does not flip to false during update (sort of indicating a change) - I don't feel strongly that it should - it will not fulfil CG use case to indicate that all replicas are now serving the new spec.

That can be OK as long as there are no vetos to this. In such case:

- we can narrow down this issue to make a decision what particular conditions should `Ready` condition look after
- create a separate issue for:
  - specifying and creating a new condition which will report the up to date state of the underlying replicas ( `RolloutComplete` or not) 
  - or document users how to go through the ownership chain and deduce this from the managed `Deployment`'s `status.updatedReplicas` field

**Comment by rainest** [comment]
tl;dr
- Use the fields, not conditions, for populating red/green lights on the dashboard. They're always the current state; conditions are past state transitions that don't fully represent the current state.
- Probably avoid an aggregate `Ready` condition altogether. DataPlane manages a set of several resources that each have their own multidimensional statuses and developing an intuitive notion of "Ready" across all of those is quite difficult. You could maybe aggregate them into a single status in an overview dashboard, but the Kubernetes statuses should be finer-grained.
- Consider narrowing the scope of what DataPlane reports when status information is available from child resources. Deployment readiness is already available in the Deployment itself, and pointing to it rather than mirroring data is a reasonable approach.

I don't think conditions are the best choice for the type of task that prompted this originally. If you're showing the current status of the DataPlane in a dashboard, that information should be sourced from the dedicated status fields (`readyReplicas` and such), not the conditions. The conditions are more to show state transition history and not so much provide rich information about the current resource status.

That said, a single boolean `Ready` status (even if there was such a status field) can't really represent the state of a Deployment well. Should you (a user with no intimate knowledge of the code) take that to mean that all replicas are up to date? That only some replicas are up to date, but that the minimum replica count meets the requested count? That all replicas are up to date, but not enough are ready to meet the minimum requested? A boolean Ready/not Ready state can't really represent several per-replica states intuitively. Conditions should represent the different tasks the DataPlane resource has separately, i.e. have separate MinReplicasAvailable and AllResourcesUpToDate conditions.

Ready as an aggregate status across the different managed resources is further confusing: the DataPlane spawns both a Deployment and several Services. Should Ready be `true` if the Deployment has minimum replicas, but the proxy Service hasn't received an LB IP allocation? At present we do, but you could well argue that it should be `false`. In either case, Ready is overloaded with many possible `false` states that need very different fixes.

The total replicas, updated replicas, ready replicas, and available replicas counts do mean different things, so if we're going to nest Deployment status under DataPlane status, we should probably include all of them. We do already include some, but omit others.

We may want to reconsider whether we should include child resource status information in the DataPlane status at all. The DataPlane spawns multiple resources--including Deployments, Services, Certificates, etc.--each with their own statuses. Currently our status fields are kind of another (partial) view of the Deployment and Service statuses, but we aren't actually the controller that manages those. IMO it'd make sense to focus the DataPlane status on the task it's assigned to (keeping resources up to date or reporting why updates are blocked) and provide links to the other resources. DataPlane status information would be limited, but dashboards could perform lookups for all the managed resources and display their statuses alongside each other. This is roughly how GKE handles Deployments; it displays a list of ReplicaSet and Pod statuses in subsections.

That said, mirroring child statuses isn't unusual--Deployment statuses are already basically summed copies of ReplicaSet statuses with the addition of `updatedReplicas` tallied only from the latest one:

```
// existing ReplicaSet getting drained as replacements come online

{
  "availableReplicas": 3,
  "fullyLabeledReplicas": 3,
  "observedGeneration": 5,
  "readyReplicas": 3,
  "replicas": 3
}

// new ReplicaSet
{
  "availableReplicas": 1,
  "fullyLabeledReplicas": 2,
  "observedGeneration": 2,
  "readyReplicas": 1,
  "replicas": 2
}

// Deployment status

{
  "availableReplicas": 4, // 3+1
  "observedGeneration": 8,
  "readyReplicas": 4, // 3+1
  "replicas": 5, // 3+2
  "unavailableReplicas": 1,
  "updatedReplicas": 2
}
```
DataPlanes manage more varied resources than Deployments (which only manage ReplicaSets) do, but you could reasonably argue that the Deployment and proxy Service are important enough to mirror their statuses. Mirroring _everything_ seems ill-advised though.

**Comment by pmalek** [comment]
> Use the fields, not conditions, for populating red/green lights on the dashboard. They're always the current state; conditions are past state transitions that don't fully represent the current state.

I see the desire for simplification but status conditions should also describe the current state: they do possess `lastTransitionTime` but they should express the current state of things with a bit more context than simple fields (we've already had an issue with simply using fields in CG use case, can't find a thread for this now though).

> Probably avoid an aggregate Ready condition altogether. DataPlane manages a set of several resources that each have their own multidimensional statuses and developing an intuitive notion of "Ready" across all of those is quite difficult. You could maybe aggregate them into a single status in an overview dashboard, but the Kubernetes statuses should be finer-grained.

I'd try to go this route, similarly as it's done with `Gateway` status, which has e.g. `GatewayService` and `DataPlaneReady` conditions which inform about each managed resource or aspect of the Gateway independently. `Ready` status condition could also do the same as `Gateway` does: set it to true when all other conditions are true. 

My rough proposal is in : https://github.com/Kong/gateway-operator/pull/1348#issuecomment-1900395482, found when working on the `Ready` status condition flapping.

> Consider narrowing the scope of what DataPlane reports when status information is available from child resources. Deployment readiness is already available in the Deployment itself, and pointing to it rather than mirroring data is a reasonable approach.

I'd argue that we should be as user friendly as possible with this, thus trying to inform about all the managed resources of `DataPlane`. Take `Deployment` as a good example of that: it doesn't tell you to go look through `ReplicaSet`s to figure out if it's ready or not, it tells you that straight in its status which is convenient and I believe we (`DataPlane`s) should be too.

**Comment by sandromodarelli** [comment]
@justindavies JFYI here's the discussion about how we should set a DataPlane ready looking at replicas. during the release of KO 2.0.1 we applied the ready condition on "at least one ready replica" to have a ready dataplane. let me know you feedback and if everything is already set and in place from your point of view, we can close this one

**Comment by sandromodarelli** [comment]
closing this because fixed by #2291

**Comment by pmalek** [comment]
> during the release of KO 2.0.1 we applied the ready condition on "at least one ready replica" to have a ready dataplane.

Just to be a bit more precise on the context of this: we currently set the `DataPlane` as not ready if it:
- either has 0 replicas
- or `AvailableReplicas` is less than `Replicas`

Source: https://github.com/Kong/kong-operator/blob/42b1cbbee9568d171ef29d4ddc281442156a5190/controller/dataplane/controller_utils.go#L303-L306

---

This issue was aiming to provide more detailed `DataPlane` deployment (not `Deployment`) status in general.

@sentinelleader @pbaderia01 I'm happy to sync again on this if there is still some information that is missing from your PoV for CGW or serverless that we could add. LMK
